lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpLink.java

```diff
@@ -0,0 +1,138 @@
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelFuture;
+import io.netty.handler.codec.http.*;
+import org.apache.reef.wake.remote.Encoder;
+import org.apache.reef.wake.remote.transport.Link;
+import org.apache.reef.wake.remote.transport.LinkListener;
+
+import java.net.SocketAddress;
+import java.net.URI;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Link implementation with Netty.
+ *
+ * If you set a {@code LinkListener<T>}, it keeps message until writeAndFlush operation completes
+ * and notifies whether the sent message transferred successfully through the listener.
+ */
+public final class NettyHttpLink<T> implements Link<T> {
+
+  public static final int INT_SIZE = Integer.SIZE / Byte.SIZE;
+
+  private static final Logger LOG = Logger.getLogger(NettyHttpLink.class.getName());
+
+  private final Channel channel;
+  private final Encoder<? super T> encoder;
+  private final LinkListener<? super T> listener;
+  private final URI uri;
+
+  /**
+   * Constructs a link.
+   *
+   * @param channel the channel
+   * @param encoder the encoder
+   * @param uri     the URI
+   */
+  public NettyHttpLink(final Channel channel,
+                       final Encoder<? super T> encoder,
+                       final URI uri) {
+    this(channel, encoder, null, uri);
+  }
+
+  /**
+   * Constructs a link.
+   *
+   * @param channel  the channel
+   * @param encoder  the encoder
+   * @param listener the link listener
+   * @param uri      the URI
+   */
+  public NettyHttpLink(
+          final Channel channel,
+          final Encoder<? super T> encoder,
+          final LinkListener<? super T> listener,
+          final URI uri) {
+    this.channel = channel;
+    this.encoder = encoder;
+    this.listener = listener;
+    this.uri = uri;
+  }
+  /**
+   * Writes the message to this link.
+   *
+   * @param message the message
+   */
+  @Override
+  public void write(final T message) {
+    LOG.log(Level.FINEST, "write {0} :: {1}", new Object[] {channel, message});
+    try {
+      final FullHttpRequest request =
+          new DefaultFullHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.POST, uri.getRawPath());
+      final ByteBuf buf = Unpooled.copiedBuffer(encoder.encode(message));
+      request.headers()
+          .set(HttpHeaders.Names.HOST, uri.getHost())
+          .set(HttpHeaders.Names.CONNECTION, HttpHeaders.Values.KEEP_ALIVE)
+          .set(HttpHeaders.Names.ACCEPT_ENCODING, HttpHeaders.Values.GZIP)
+          .set(HttpHeaders.Names.CONTENT_TYPE, "application/wake-transport")
+          .set(HttpHeaders.Names.CONTENT_LENGTH, buf.readableBytes());
+      request.content().clear().writeBytes(buf);
+      final ChannelFuture future = channel.writeAndFlush(request);
+      if (listener != null) {
+        future.addListener(new NettyChannelFutureListener<>(message, listener));
+      }
+      future.sync();
```

Comment:
It is because sync() will throw an Exception when sending request is Interrupted. To catch a failure on sending request, sync was used here.

Reply:
I mean, why do we need sync? you already registered the future listener and it doesn't look we need sync()

Reply:
`listener` can be null in constructor. If no listener's are attached, there would be no warning about failure on sending request. In my opinion, if default listener for failure detect is provided, we don't have to use `sync()`.

Reply:
Using `Future` and `listener` is to asynchronously write the requests, but if you do `sync()` immediately, I think there is no benefit of the asynchronous write.

Reply:
I stronly agree with you. Then what about change the implmentation to adopt default listener for detect failure on complete? or just delete `sync()`? Can I ask which would be better choice?

Reply:
Sorry, I don't fully understand your question. You've already registered the future listener in line 104.

Reply:
@taegeonum the `listener` can be null. The constructor of `NettyHttpLink` at line59 makes listener null. So There can be no listener for catching failure if `sync()` is not performed.

Reply:
@nhne I see. I think it is okay to me. If we want to ignore the message futures, we can skip the listener. Otherwise, we provide the listener and handle the message futures. So, how about just removing `sync()`?

Reply:
Then I'll remove `sync()`. User may provide listener if they need any failure detection. Thank you for sharing your opinion!